lighthouse-core/lib/client-rect-functions.js

```diff
@@ -0,0 +1,311 @@
+  );
+}
+
+/**
+ * Merge client rects together. This may result in a larger overall size than that of the individual client rects.
+ * @param {LH.Artifacts.ClientRect[]} clientRects
+ */
+function simplifyClientRects(clientRects) {
+  clientRects = filterOutTinyClientRects(clientRects);
+  clientRects = filterOutClientRectsContainedByOthers(clientRects);
+  clientRects = mergeTouchingClientRects(clientRects);
+  return clientRects;
+}
+
+/**
+ * @param {LH.Artifacts.ClientRect[]} clientRects
+ * @returns {LH.Artifacts.ClientRect[]}
+ */
+function filterOutTinyClientRects(clientRects) {
+  // 1x1px rect shouldn't be reason to treat the rect as something the user should tap on.
+  // Often they're made invisble in some obscure way anyway, and only exist for e.g. accessibiliity.
+  const nonTinyClientRects = clientRects.filter(
+    rect => rect.width > 1 && rect.height > 1
+  );
+  if (nonTinyClientRects.length > 0) {
+    return nonTinyClientRects;
+  }
+  return clientRects;
+}
+
+const rectContainsString = `
+  ${rectContainsPoint.toString()}
+  ${rectContains.toString()};
+`;
+
+/**
+ * @param {LH.Artifacts.ClientRect[]} clientRects
+ * @returns {LH.Artifacts.ClientRect[]}
+ */
+function filterOutClientRectsContainedByOthers(clientRects) {
+  const filteredOutRects = new Set();
+  return clientRects.filter(cr => {
+    for (const possiblyContainingRect of clientRects) {
+      if (possiblyContainingRect === cr) {
+        continue;
+      }
+      if (filteredOutRects.has(possiblyContainingRect)) {
+        // Can't contain anything because we removed it from the list
+        continue;
+      }
+      if (rectContains(possiblyContainingRect, cr)) {
+        filteredOutRects.add(cr);
+        return false;
+      }
+    }
+    return true;
+  });
+}
+
+/**
+ * @param {number} a
+ * @param {number} b
+ */
+function almostEqual(a, b) {
+  // Sometimes a child will reach out of the parent by
+  // 1px or 2px, so be somewhat tolerant for merging
+  return Math.abs(a - b) <= 2;
+}
+
+/**
+ * @param {LH.Artifacts.ClientRect} rect
+ */
+function getRectCenterPoint(rect) {
+  return {
+    x: rect.left + rect.width / 2,
+    y: rect.top + rect.height / 2,
+  };
+}
+
+/**
+ * @param {LH.Artifacts.ClientRect} crA
+ * @param {LH.Artifacts.ClientRect} crB
+ * @returns {boolean}
+ */
+function clientRectsTouch(crA, crB) {
+  // https://stackoverflow.com/questions/2752349/fast-rectangle-to-rectangle-intersection
+  return (
+    crA.left <= crB.right &&
+    crB.left <= crA.right &&
+    crA.top <= crB.bottom &&
+    crB.top <= crA.bottom
+  );
+}
+
+/**
+ * @param {LH.Artifacts.ClientRect[]} clientRects
+ * @returns {LH.Artifacts.ClientRect[]}
+ */
+function mergeTouchingClientRects(clientRects) {
+  for (let i = 0; i < clientRects.length; i++) {
+    for (let j = i + 1; j < clientRects.length; j++) {
+      const crA = clientRects[i];
+      const crB = clientRects[j];
+
+      /**
+       * Examples of what we want to merge:
```

Comment:
Maybe a brief comment on why we wouldn't merge them if they overlap but don't align? perhaps this belongs on what simplify actually does, but it's not clear from this alone why it shouldn't merge them

Reply:
We want to merge two rects if the user would perceive them as a single tappable area to tap on. The user would intuitively tap on the center of the merged rect – we try to determine that rect and then place the simulated finger area in the middle.

Tried to make this a bit clearer in the code.

Reply:
Ah gotcha, thanks!